### PR TITLE
add new command kubectl set volume

### PR DESCRIPTION
**Pull request by WanLinghao**
**What this PR does / why we need it**:
#21648 
Moved from OpenShift to Kubenetes.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```
Add 'kubectl set volume' command for setting volume in containers for resources embedding pod templates
```

### CODE REVIEW

**Comment by mkumatag** [comment]
`api.Codecs` moved to `legacyscheme.Codecs` part of https://github.com/kubernetes/kubernetes/pull/53984, so you may need to fix the import path to build the code.

**Comment by cblecker** [comment]
/unassign
Please reassign back if you need `hack/` approval once this has been reviewed by a sig-cli reviewer :)

**Comment by kargakis** [comment]
@kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/set/set_volume.go
Comment (dixudx) [inline pkg/kubectl/cmd/set/set_volume.go]: Better change to

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

Comment (dixudx) [inline pkg/kubectl/cmd/set/set_volume.go]: Better change to

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

Comment (dixudx) [inline pkg/kubectl/cmd/set/set_volume.go]: Better change to 

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

File: pkg/kubectl/cmd/set/set_volume.go
Comment (dixudx) [inline pkg/kubectl/cmd/set/set_volume.go]: I wonder shall we split these checking to separate methods to make it more concise and maintainable.

File: pkg/kubectl/cmd/set/set_volume.go
Comment (dixudx) [inline pkg/kubectl/cmd/set/set_volume.go]: Actually I don't like such a hard-coded way.

Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: @dixudx please make it more clear.Do you mean this function contains too much things?

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: I think cause o.Name is an existed variable, so your way is not suitable

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: @dixudx OK

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]:
```
var err error
if o.Name, err = o.getVolumeName(spec, singleResource); err != nil {
	return err
}

```
did you mean this?

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: OK

File: pkg/kubectl/cmd/set/set_volume.go
Comment (dixudx) [reply pkg/kubectl/cmd/set/set_volume.go]: SGTM

File: pkg/kubectl/cmd/set/set_volume.go
Comment (shiywang) [inline pkg/kubectl/cmd/set/set_volume.go]: no internal type

**Review by shiywang — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/set/set_volume.go
Comment (shiywang) [inline pkg/kubectl/cmd/set/set_volume.go]: nit: i would prefer just "k8s.io/apimachinery/pkg/api/resource" since there's no upstream dependency anymore.

File: pkg/kubectl/cmd/set/set_volume.go
Comment (shiywang) [inline pkg/kubectl/cmd/set/set_volume.go]: in kubernetes, we can just call it client

File: pkg/kubectl/cmd/set/set_volume.go
Comment (shiywang) [inline pkg/kubectl/cmd/set/set_volume.go]: ditto

**Comment by shiywang** [comment]
just have a glance, will take a deeper look tomorrow morning

File: pkg/kubectl/cmd/set/set_volume_test.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume_test.go]: this should be removed, use versioned api

File: pkg/kubectl/cmd/set/set_volume_test.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume_test.go]: ditto

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: should be versioned

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: this format is on purpose?

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: should be use kubernetes.IO not openshift

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]:
```
		// if a --local flag was provided, and a resource was specified in the form
		// <resource>/<name>, fail immediately as --local cannot query the api server
		// for the specified resource.
		if len(o.Resources) > 0 {
			return resource.LocalResourceError
		}
```
should be check when --local and resources both special

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: use build will better instead b

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: if multi resources get updated, this will print just one

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: should be use `utilerrors.NewAggregate(allErrs)` will better

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: since this file import another resource package   "k8s.io/kubernetes/pkg/kubectl/resource"

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: this format was imported from original file, I will keep it  same with k8s style

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: what about just delete this line

File: pkg/kubectl/cmd/set/set_volume_test.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume_test.go]: use "k8s.io/client-go/kubernetes/typed/core/v1" replacing  "k8s.io/kubernetes/pkg/api"?

Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume_test.go]: @zjj2wry since I am still not clear with this version stuff

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: @shiywang  I am not clear with version stuff,  would please give me some hints, thanks a lot

File: pkg/kubectl/cmd/set/set_volume_test.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume_test.go]: api rep: `k8s.io/api...`
api maybe in: `k8s.io/api/core/v1`

Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume_test.go]: @zjj2wry  Is it a new move still on pushing? I notice that file in same package like set_env.go, set_image.go still use "k8s.io/kubernetes/pkg/api". 
since connections between them exist, if use core/v1 in set_volume.go. then it is necessary that  we should use the same in this whole package

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume_test.go]: yes, will fix all set command :), and it is in process

Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume_test.go]: @zjj2wry so would mind if I just ignore this right now and fix it after other set command?
by the way , are there any PR that focus on this? please tell me so I can keep in track

**Comment by WanLinghao** [comment]
@cblecker PTAL

File: pkg/kubectl/cmd/set/set_volume_test.go
Comment (liggitt) [reply pkg/kubectl/cmd/set/set_volume_test.go]: it's being actively worked on in https://github.com/kubernetes/kubernetes/pull/54554

@WanLinghao I would prefer not adding this until that PR is in

Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume_test.go]: @liggitt  OK, I will wait for this and rebase then

**Comment by WanLinghao** [comment]
#54554 rebase after this patch

File: pkg/kubectl/cmd/set/set_volume.go
Comment (WanLinghao) [inline pkg/kubectl/cmd/set/set_volume.go]: @zjj2wry @liggitt I got trouble here, can't find Kind function in "k8s.io/api/core/v1"

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume.go]: GroupVersionKind.GroupKind()

Comment (WanLinghao) [reply pkg/kubectl/cmd/set/set_volume.go]: @zjj2wry can you describe it in detail?thanks a lot

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume.go]: `mapping, err := mapper.RESTMapping(schema.GroupKind{Kind: "PersistentVolumeClaim"})` maybe can work

**Comment by WanLinghao** [comment]
@zjj2wry @kargakis @cblecker  reabased, PTAL

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: nits，use coreclient

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: what about apiresource

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: project is openshift define, should update it to namespace

File: pkg/kubectl/cmd/set/set_volume.go
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_volume.go]: The `--remove` functionality will likely end up in its own [`kubectl unset`](https://github.com/kubernetes/kubernetes/issues/39775) command and `--list` in [`kubectl view`](https://github.com/kubernetes/kubernetes/issues/29679). That being said,I would suggest making `--add` the default here and remove all of these flags. 

cc: @kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: use o.Mapper direct

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume.go]: +1

File: pkg/kubectl/cmd/set/set_volume.go
Comment (zjj2wry) [inline pkg/kubectl/cmd/set/set_volume.go]: this maybe will occur error

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume.go]: https://github.com/kubernetes/kubernetes/issues/55407

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_volume.go]: and @smarterclayton doing refactor..

https://github.com/kubernetes/kubernetes/pull/55660#issuecomment-344172145

**Comment by zjj2wry** [comment]
@WanLinghao this PR lgtm when your fix small nits.

**Comment by WanLinghao** [comment]
@zjj2wry  @kargakis  ready

**Comment by WanLinghao** [comment]
@liggitt PTAL

**Comment by WanLinghao** [comment]
@zjj2wry   do you know who can help push this, thanks

**Comment by WanLinghao** [comment]
@liggitt  PTAL  thanks

**Comment by zjj2wry** [comment]
/retest
@kubernetes/sig-cli-pr-reviews

**Comment by mengqiy** [comment]
Haven't look at the code closely, but I'm happy this PR doesn't introduce bad dependency.

**Comment by WanLinghao** [comment]
@liggitt  are you ok with this patch? if so , when it  can be merged.

**Comment by dixudx** [comment]
@WanLinghao Code freeze now for v1.9 release. And this is not a must-fix issue for v1.9.

Since it has already been labeled for v1.10, please be patient to wait for the end of release v1.9.

Thanks for your contribution.

**Comment by WanLinghao** [comment]
@dixudx OK,thanks

**Comment by WanLinghao** [comment]
@smarterclayton @liggitt PTAL

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/54284)

**Pull request closed without merging**
